TS/JS language support; Windows support

alexsaveliev opened the pull request:
- Windows support (dealing with Windows-specific URI `file:///`, looking for LSP server executable in `PATH` while trying extensions from `PATHEXT`)
- Decorating results of "Find References" by using local file system data in a separate thread. We need to have some data not provided by the LSP protocol (such as matching lines text and symbol information) to construct "Find References" results page
- Added installation instructions for TypeScript/JavaScript
- Dumping LSP server's stderr messages

alexsaveliev committed:
WIP

- Windows support
- Some logging
- Installation instructions (WIP)

alexsaveliev committed:
Windows support (WIP)

alexsaveliev committed:
Extracting data to display in "find references" panel

- extracting data to display in "find references" panel. Reading files from FS and extracting Nth line using reference location spans. Caching files read and line offsets
- using symbol name as 'symbol display string' to avoid 'quickinfo' requests

alexsaveliev committed:
Printing content of LSP server stderr

alexsaveliev committed:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	README.md

alexsaveliev committed:
Formatting

alexsaveliev committed:
Decorating references in a separate thread

- to unblock message receiving thread, decorating references (by adding `lineText` attribute) in a separate thread

uforic commented on typescript/libs/lsp_client.py:
```diff
@@ -243,31 +293,37 @@ def __init__(self, binary_path, args, env, root_path):
             readerThread.daemon = True
             readerThread.start()
-        self.postCmd(lsp_helpers.init_message("file://" + self.root_path, self.server_proc.pid))
+            log.debug("starting logger thread")
```
On my machine, this error logger froze Sublime until I commented it out. I can't figure out why - any thoughts?

alexsaveliev replied:
Should be addressed by https://github.com/alexsaveliev/sublime-lsp/commit/7c809777a5727fefd58b53d6eebaacaaf7948b9a#diff-5cde716076cda2c46cc19898b623da71R334

uforic commented on typescript/libs/lsp_client.py:
```diff
@@ -26,7 +27,10 @@ class LspCommClient(node_client.CommClient):
         self.server_proc = None
         self.args = copy.deepcopy(args)
-        self.args.insert(0, binary_path)
+        program = LspCommClient.which(binary_path)
```
What does this get us? Why bother calling which, as opposed to having the user hardcode in the address where the binary is (which they'll probably have to do anyways since Sublime Text 3 doesn't capture paths set up in .bashrc?

The which command below is quite a lot of code!

alexsaveliev replied:
> What does this get us?

Ability to search for proper executable in Windows `PATH` env using "executable" extensions defined in `PATHEXT` env. For example use can put `javascript-typescript-stdio` as `binary` and then extension will transform it to '%JSTSDIR%\bin\javascript-typescript-stdio.bat` on Windows

> The which command below is quite a lot of code!

Well, it's a copy of `shutil.which` introduced in Python 3.3 :) I am not sure if we can rely on Python 3.3 presence so added it

uforic commented on typescript/commands/references.py:
```diff
@@ -173,3 +191,42 @@ def run(self, text, argsJson):
+        return content[start:end].rstrip()
+
+    def __table(self, fileName):
+        """ returns line table for a given filename, computes if needed """
```
This should be in **init**, right? Any reason to do the initialization lazily?

alexsaveliev replied:
> This should be in init, right? Any reason to do the initialization lazily?

Nope, I think that we should initialize line offset tables lazily because purpose of this class is to transform pairs `(file F, line L)` into `text of line L in file F`. I see no reason to iterate over all references and extract unique file names before asking for line's text just to move this code to `init`

uforic commented on typescript/commands/references.py:
```diff
@@ -173,3 +191,42 @@ def run(self, text, argsJson):
+        if lines is not None:
+            return lines
+        f = open(fileName)
+        lines = []
```
Wouldn't it be better to collect the lines as we read them out of the file? With this way, we are reading the entire file into memory, whereas if we iterate one line at a time and toss the lines when we don't need them, we only ever have a line in memory

alexsaveliev replied:
> Wouldn't it be better to collect the lines as we read them out of the file? 

There may be multiple references to symbol in the same file. We may reduce I/O operations if we'll read and cache them

uforic commented on README.md:
````diff
@@ -62,13 +81,70 @@ Next, configure the LSP connector for the `langserver-go` binary. To change your
+
+```
+{
+...
````
Any reason to separate these out? It ends up looking more verbose I think. 

```
{
clients:[
...
{}
...
]

}
```

alexsaveliev replied:
Fixed

uforic commented on README.md:
````diff
@@ -62,13 +81,70 @@ Next, configure the LSP connector for the `langserver-go` binary. To change your
+```
+{
+        {
+            "binary": "javascript-typescript-sublime",
````
Do we need to call it -sublime, or is it actually just a standard-in standard-out version? "javascript-typescript-stdio" or something?

alexsaveliev replied:
Renamed to `javascript-typescript-stdio`

uforic reviewed (changes requested):
Hey! looking great - I couldn't get it working for me on TypeScript, on the sublime version of js-ts-server on your account, I was getting these errors:
"Content-Length: 98

{"jsonrpc":"2.0","id":1,"error":{"code":-32601,"message":"Unhandled method textDocument/didOpen"}}"

uforic commented on typescript/commands/references.py:
```diff
@@ -173,3 +191,42 @@ def run(self, text, argsJson):
+        lines = self.__lines.get(fileName)
+        if lines is not None:
+            return lines
+        f = open(fileName)
```
Use utf-8 encoding, on my machien I'm getting errors in this method:
`UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 2218: ordinal not in range(128)`

uforic replied:
Ideally, we wouldn't have to guess the encoding, alternatively.

alexsaveliev replied:
Done, using `utf-8` and ignoring decoding errors

uforic commented on typescript/commands/references.py:
```diff
@@ -36,6 +37,23 @@ def is_visible(self):
 
+    @staticmethod
+    def __decorate(file_name, line, references):
+        helper = RefsHelper()
```
Error handling - not being able to open a file or some weird IO error will be pretty common, so let's just fill the text with "Error retrieving preview," and successfully parse the rest

alexsaveliev replied:
Done

alexsaveliev commented:
> I couldn't get it working for me on TypeScript, on the sublime version of js-ts-server on your account, I was getting these errors:

Could you please provide full log?

alexsaveliev committed:
Addressed code review comments:

- handling cases when we were unable to read file to fetch reference's line text: in this situation we'll display `Error fetching preview from ...` as line content
- when reading files to extract reference's line text, using `utf-8` encoding and ignoring decoding errors
- closing stderr pipe when LSP server exited to terminate logger thread
- updated README, removed unnecessary verbose blocks

Commit:
lsp: combine file seeks and only take lines we need

Commit:
remove refhelper call

uforic reviewed (approved):
LGTM! made a small change to unblock this, not using an object to read a file but a method to grab relevant lines for a file. This avoids pinning the whole file in memory. Will merge now, thx!
